core/lib/types/src/storage_writes_deduplicator.rs

```diff
@@ -167,17 +180,33 @@ impl StorageWritesDeduplicator {
                 &mut self.metrics.repeated_storage_writes
             };
 
+            let total_size = &mut self.metrics.total_updated_values_size;
+
             match item.update_type {
                 UpdateType::Insert => {
-                    self.modified_key_values.remove(&item.key);
+                    let value = self
+                        .modified_key_values
+                        .remove(&item.key)
+                        .unwrap_or_else(|| {
+                            panic!("tried removing key: {:?} before insertion", item.key)
+                        });
                     *field_to_change -= 1;
+                    *total_size -= value.compressed_size;
                 }
                 UpdateType::Update(value) => {
-                    self.modified_key_values.insert(item.key, value);
+                    let old_value = self
+                        .modified_key_values
+                        .insert(item.key, value)
+                        .unwrap_or_else(|| {
+                            panic!("tried removing key: {:?} before insertion", item.key)
+                        });
+                    *total_size += value.compressed_size;
+                    *total_size -= old_value.compressed_size;
                 }
                 UpdateType::Remove(value) => {
                     self.modified_key_values.insert(item.key, value);
                     *field_to_change += 1;
+                    *total_size += value.compressed_size;
```

Comment:
ah oh this method is complex, and we adding more mutable/imperative statements.. I don't see a quick improvement here - just something to keep in mind in the future